Refactor kubelet config validation tests

shuheiktgw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/sig node
/kind cleanup
/area test

#### What this PR does / why we need it:
The background is described in #105029. I also noticed several minor problems in the validations so I fixed them:
- Missing the `invalid validation:` prefix 
- Not showing a flag in an error message. For example `CPUCFSQuotaPeriod`
- Not showing a given value in an error message. For example `NodeStatusMaxImages`
- Not showing possible options in an error message. For example `TopologyManagerPolicy`
- Unnecessary capitalization. For example `ShutdownGracePeriod`
- Use a case statement to validate options. For example `MemorySwap.SwapBehavior`
- Use `%q` instead of `%v` if necessary. For example `EnforceNodeAllocatable`
- And a few other minor fixes (I'll leave a comment on them)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #105029

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubelet config validation error messages are updated
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
None
```

shuheiktgw commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -217,7 +224,7 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	if kc.MemoryThrottlingFactor != nil && (*kc.MemoryThrottlingFactor <= 0 || *kc.MemoryThrottlingFactor > 1.0) {
-		allErrors = append(allErrors, fmt.Errorf("invalid configuration: memoryThrottlingFactor %v must be greater than 0 and less than or equal to 1.0", kc.MemoryThrottlingFactor))
+		allErrors = append(allErrors, fmt.Errorf("invalid configuration: memoryThrottlingFactor %v must be greater than 0 and less than or equal to 1.0", *kc.MemoryThrottlingFactor))
```
I'm pretty sure a used doesn't want to see a memory location 😄

shuheiktgw commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -195,10 +202,10 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 		if _, err := cpuset.Parse(kc.ReservedSystemCPUs); err != nil {
-			allErrors = append(allErrors, fmt.Errorf("unable to parse reservedSystemCPUs (--reserved-cpus), error: %v", err))
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: unable to parse reservedSystemCPUs (--reserved-cpus) %v, error: %w", kc.ReservedSystemCPUs, err))
 		}
```
Better to use `%w` for an error

shuheiktgw commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
+		if kc.ShutdownGracePeriodCriticalPods.Duration > kc.ShutdownGracePeriod.Duration {
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: shutdownGracePeriodCriticalPods %v must be <= ShutdownGracePeriod %v", kc.ShutdownGracePeriodCriticalPods, kc.ShutdownGracePeriod))
 		}
-		if kc.ShutdownGracePeriod.Duration > 0 && kc.ShutdownGracePeriod.Duration < time.Duration(time.Second) {
```
Casting `time.Second` to `time.Duration` is unnecessary

shuheiktgw commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 
 	if localFeatureGate.Enabled(features.GracefulNodeShutdown) {
-		if kc.ShutdownGracePeriod.Duration < 0 || kc.ShutdownGracePeriodCriticalPods.Duration < 0 || kc.ShutdownGracePeriodCriticalPods.Duration > kc.ShutdownGracePeriod.Duration {
```
Merged this validation with the later one to simplify the logic

pacoxu replied:
/cc @wzshiming

wzshiming replied:
/cc @bobbypage

pacoxu commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -14,22 +14,25 @@ See the License for the specific language governing permissions and
 */
 
-package validation
+package validation_test
```
s/validation_test/validation/g

shuheiktgw replied:
I thought that was one of the best practices to remove test files whenever we don't have to call unexported methods/functions from a test file but no? I found some test files with the `_test` suffix in this project but it does not seem that popular though...

Ref: https://pkg.go.dev/cmd/go@master#hdr-Test_packages
> Test files that declare a package with the suffix "_test" will be compiled as a separate package, and then linked and run with the main test binary.

ehashman replied:
Kubernetes doesn't necessarily follow all golang conventions, we do this very inconsistently in kubelet:

```
ehashman@fedora:~/src/k8s$ git grep 'package .\+_test' pkg/kubelet/
pkg/kubelet/container/runtime_cache_test.go:package container_test
pkg/kubelet/envvars/envvars_test.go:package envvars_test
pkg/kubelet/kubelet_dockerless_test.go:package kubelet_test
pkg/kubelet/runtimeclass/runtimeclass_manager_test.go:package runtimeclass_test
```

We can leave this as is or not, I don't think it really matters.

pacoxu commented:
/test pull-kubernetes-unit

shuheiktgw commented:
Whoops, I haven't noticed there are already tests for the ReservedMemory configurations, thanks! I removed the duplicate tests and updated existing ones 👍

shuheiktgw commented:
@pacoxu Sorry I forgot to reassign this PR to you after I fixed it! Would you review the PR again? 🙏

shuheiktgw commented:
/retest

pacoxu reviewed (approved):
The validation error messages are more clear and unified.
/lgtm

wzshiming commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
-		if kc.ShutdownGracePeriod.Duration < 0 || kc.ShutdownGracePeriodCriticalPods.Duration < 0 || kc.ShutdownGracePeriodCriticalPods.Duration > kc.ShutdownGracePeriod.Duration {
-			allErrors = append(allErrors, fmt.Errorf("invalid configuration: ShutdownGracePeriod %v must be >= 0, ShutdownGracePeriodCriticalPods %v must be >= 0, and ShutdownGracePeriodCriticalPods %v must be <= ShutdownGracePeriod %v", kc.ShutdownGracePeriod, kc.ShutdownGracePeriodCriticalPods, kc.ShutdownGracePeriodCriticalPods, kc.ShutdownGracePeriod))
+		if kc.ShutdownGracePeriodCriticalPods.Duration > kc.ShutdownGracePeriod.Duration {
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: shutdownGracePeriodCriticalPods %v must be <= ShutdownGracePeriod %v", kc.ShutdownGracePeriodCriticalPods, kc.ShutdownGracePeriod))
```
```suggestion
			allErrors = append(allErrors, fmt.Errorf("invalid configuration: shutdownGracePeriodCriticalPods %v must be <= shutdownGracePeriod %v", kc.ShutdownGracePeriodCriticalPods, kc.ShutdownGracePeriod))
```

wzshiming commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	if (kc.ShutdownGracePeriod.Duration > 0 || kc.ShutdownGracePeriodCriticalPods.Duration > 0) && !localFeatureGate.Enabled(features.GracefulNodeShutdown) {
-		allErrors = append(allErrors, fmt.Errorf("invalid configuration: Specifying ShutdownGracePeriod or ShutdownGracePeriodCriticalPods requires feature gate GracefulNodeShutdown"))
+		allErrors = append(allErrors, fmt.Errorf("invalid configuration: specifying ShutdownGracePeriod or ShutdownGracePeriodCriticalPods requires feature gate GracefulNodeShutdown"))
```
```suggestion
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: specifying shutdownGracePeriod or shutdownGracePeriodCriticalPods requires feature gate GracefulNodeShutdown"))
```

wzshiming commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
+		case kubetypes.LimitedSwap:
+		case kubetypes.UnlimitedSwap:
+		default:
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: memorySwap.SwapBehavior %q must be one of: \"\", %q, or %q", kc.MemorySwap.SwapBehavior, kubetypes.LimitedSwap, kubetypes.UnlimitedSwap))
```
```suggestion
			allErrors = append(allErrors, fmt.Errorf("invalid configuration: memorySwap.swapBehavior %q must be one of: \"\", %q, or %q", kc.MemorySwap.SwapBehavior, kubetypes.LimitedSwap, kubetypes.UnlimitedSwap))
```

wzshiming commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	if !localFeatureGate.Enabled(features.NodeSwap) && kc.MemorySwap != (kubeletconfig.MemorySwapConfiguration{}) {
-		allErrors = append(allErrors, fmt.Errorf("invalid configuration: MemorySwap.SwapBehavior cannot be set when NodeSwap feature flag is disabled"))
+		allErrors = append(allErrors, fmt.Errorf("invalid configuration: memorySwap.SwapBehavior cannot be set when NodeSwap feature flag is disabled"))
```
```suggestion
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: memorySwap.swapBehavior cannot be set when NodeSwap feature flag is disabled"))
```

wzshiming reviewed (commented):
Some initials are not uniform in lowercase, others look good to me

wzshiming commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

ehashman commented:
/assign

ehashman commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -132,52 +132,59 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	case kubeletconfig.SingleNumaNodeTopologyManagerPolicy:
 	default:
-		allErrors = append(allErrors, fmt.Errorf("invalid configuration: topologyManagerPolicy non-allowable value: %v", kc.TopologyManagerPolicy))
+		allErrors = append(allErrors, fmt.Errorf("invalid configuration: topologyManagerPolicy (--topology-manager-policy) %q must be one of: %q, %q, %q, or %q", kc.TopologyManagerPolicy, kubeletconfig.NoneTopologyManagerPolicy, kubeletconfig.BestEffortTopologyManagerPolicy, kubeletconfig.RestrictedTopologyManagerPolicy, kubeletconfig.SingleNumaNodeTopologyManagerPolicy))
```
(non-blocking) Can we pass these in as a slice so we don't have to change the format string every time we add a new policy?

ehashman commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -70,186 +71,454 @@ func TestValidateKubeletConfiguration(t *testing.T) {
-	if allErrors := ValidateKubeletConfiguration(successCase2); allErrors != nil {
-		t.Errorf("expect no errors, got %v", allErrors)
+
+	for _, tc := range successCases {
```
(non-blocking) Rather than splitting success and error cases, we could have a single loop and the success case simply has an empty errMsg.

ehashman commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -70,186 +71,454 @@ func TestValidateKubeletConfiguration(t *testing.T) {
+			},
+			errMsg: "invalid configuration: topologyManagerPolicy single-numa-node requires feature gate TopologyManager",
+		},
+		{
+			name: "use SingleNumaNodeTopologyManagerPolicy without enabling TopologyManager",
+			configure: func(conf *kubeletconfig.KubeletConfiguration) *kubeletconfig.KubeletConfiguration {
+				conf.FeatureGates = map[string]bool{"TopologyManager": false}
+				conf.TopologyManagerPolicy = kubeletconfig.SingleNumaNodeTopologyManagerPolicy
+				return conf
+			},
+			errMsg: "invalid configuration: topologyManagerPolicy single-numa-node requires feature gate TopologyManager",
+		},
```
This test case is a duplicate of the one above.

ehashman commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -70,186 +71,454 @@ func TestValidateKubeletConfiguration(t *testing.T) {
+				conf.MemoryThrottlingFactor = utilpointer.Float64(1.1)
+				return conf
+			},
+			errMsg: "invalid configuration: memoryThrottlingFactor 1.1 must be greater than 0 and less than or equal to 1.0",
```
Verified we have a correct test case corresponding to everything except these lines: https://github.com/kubernetes/kubernetes/blob/5269ac17ad5a265c10945adb95dfd5b25a0dfb29/pkg/kubelet/apis/config/validation/validation.go#L212-L221

- `validateReservedMemoryConfiguration`: has its own tests https://github.com/kubernetes/kubernetes/blob/5269ac17ad5a265c10945adb95dfd5b25a0dfb29/pkg/kubelet/apis/config/validation/validation_reserved_memory_test.go
- `validateKubeletOSConfiguration`: no existing coverage, fine
- `metrics.ValidateShowHiddenMetricsVersion`: not ours to test
- `logs.ValidateLoggingConfiguration`: not ours to test

ehashman commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -70,186 +71,454 @@ func TestValidateKubeletConfiguration(t *testing.T) {
+		t.Run(tc.name, func(t *testing.T) {
+			errs := validation.ValidateKubeletConfiguration(tc.configure(successConfig.DeepCopy()))
+			if errs == nil {
+				t.Error("expected error")
```
Maybe include what the expected error should be?

ehashman reviewed (commented):
Other than the accidentally duplicated test case, LGTM. The rest of my feedback is non-blocking but might improve this PR.

Sorry for my delay in reviewing, this was a hard one :)

This is _significantly_ improved from the current state, bravo for taking this on and writing such a nice patch!

shuheiktgw commented:
@ehashman Sorry that it took me so long to fix it! I've fixed it up so would you review the PR again? 🙏

shuheiktgw commented:
/retest

ehashman commented:
Discussed at today's SIG Node CI subproject meeting, we would like to see this merge before the end of test freeze. https://docs.google.com/document/d/1fb-ugvgdSVIkkuJ388_nhp2pBTy_4HEVg5848Xy7n5U/edit#bookmark=id.jka16r6nootq

/milestone v1.23

ehashman commented on pkg/kubelet/apis/config/validation/validation_test.go:
```diff
@@ -74,31 +74,17 @@ var (
-				t.Errorf("unexpected error: %v", errs)
-			}
-		})
-	}
```
We lost the success case, we still need to test the happy path.

shuheiktgw replied:
It's here! 👍 

https://github.com/shuheiktgw/kubernetes/blob/ce20835107927ab5a19a72b98b42808dc2babe6b/pkg/kubelet/apis/config/validation/validation_test.go#L83-L86

ehashman replied:
Ah I see it now, my diff might have been wonky

ehashman reviewed (commented):
One final comment, and then can you please squash your commits down to 1-2 total logical commits?

Commit:
Refactor Kubelet config validations

Commit:
Refactor Kubelet config validation tests

shuheiktgw commented:
@ehashman I've squashed the commits to two! 🙏

shuheiktgw commented:
/retest

ehashman reviewed (commented):
/lgtm

ehashman commented:
/assign @liggitt 

Looks like this is an API reviewers only OWNERS file

shuheiktgw commented:
Rebased the PR and resolved the conflicts

shuheiktgw commented:
/retest

ehashman reviewed (commented):
/lgtm

liggitt commented:
/approve

ehashman commented:
flake https://github.com/kubernetes/kubernetes/issues/106125
/retest-required

reylejano commented:
/milestone clear
Release Managers are cutting 1.23.0-rc.0 and creating the `release-1.23` branch

pacoxu commented:
/test pull-kubernetes-e2e-kind-ipv6

ehashman commented:
boo flakes, this will have to land when 1.24 opens :\
